engines/router/missionctl/server/grpc/server.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/caraml-dev/turing/engines/router/missionctl/log"
+	"github.com/caraml-dev/turing/engines/router/missionctl/log/resultlog"
+	"github.com/caraml-dev/turing/engines/router/missionctl/turingctx"
+	upiv1 "github.com/caraml-dev/universal-prediction-interface/gen/go/grpc/caraml/upi/v1"
+	"github.com/gojek/fiber"
+	fibergrpc "github.com/gojek/fiber/grpc"
+	"github.com/opentracing/opentracing-go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection"
+)
+
+const turingReqIDHeaderKey = "Turing-Req-ID"
+const tracingComponentID = "grpc_handler"
+
+type UPIServer struct {
+	upiv1.UnimplementedUniversalPredictionServiceServer
+
+	missionControl missionctl.MissionControlUPI
+	port           int
+}
+
+func NewUPIServer(mc missionctl.MissionControlUPI, port int) *UPIServer {
+	return &UPIServer{
+		missionControl: mc,
+		port:           port,
+	}
+}
+
+func (us *UPIServer) Run() {
+	s := grpc.NewServer()
+	upiv1.RegisterUniversalPredictionServiceServer(s, us)
+	reflection.Register(s)
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", us.port))
+	if err != nil {
+		log.Glob().Errorf("Failed to listen on port %d: %s", us.port, err)
+		return
+	}
+
+	errChan := make(chan error, 1)
+	stopChan := make(chan os.Signal, 1)
+
+	// bind OS events to the signal channel
+	signal.Notify(stopChan, syscall.SIGTERM, syscall.SIGINT)
+
+	go func() {
+		if err := s.Serve(listener); err != nil {
+			errChan <- err
+		}
+	}()
+
+	defer func() {
+		s.GracefulStop()
+	}()
+
+	// block until either OS signal, or server fatal error
+	select {
+	case err := <-errChan:
+		log.Glob().Errorf("Failed to start Turing Mission Control API: %s", err)
+	case <-stopChan:
+		log.Glob().Info("Signal to stop server")
+	}
+
+}
+
+func (us *UPIServer) PredictValues(ctx context.Context, req *upiv1.PredictValuesRequest) (
+	*upiv1.PredictValuesResponse, error) {
+	var predictionErr *errors.TuringError
+	defer metrics.GetMeasureDurationFunc(predictionErr, tracingComponentID)()
+
+	// Create context from the request context
+	ctx = turingctx.NewTuringContext(ctx)
+	// Create context logger
+	ctxLogger := log.WithContext(ctx)
+	defer func() {
+		_ = ctxLogger.Sync()
+	}()
+
+	// if request comes with metadata, attach it to metadata to be sent with fiber
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		md = metadata.New(map[string]string{})
+	}
+
+	// Get the unique turing request id from the context
+	turingReqID, err := turingctx.GetRequestID(ctx)
+	if err != nil {
+		ctxLogger.Errorf("Could not retrieve Turing Request ID from context: %v",
+			err.Error())
+	}
+	ctxLogger.Debugf("Received request for %v", turingReqID)
+	md.Append(turingReqIDHeaderKey, turingReqID)
+
+	if tracing.Glob().IsEnabled() {
+		var sp opentracing.Span
+		sp, _ = tracing.Glob().StartSpanFromContext(ctx, tracingComponentID)
+		if sp != nil {
+			defer sp.Finish()
+		}
+	}
+
+	fiberRequest := &fibergrpc.Request{
+		Message:  req,
```

Comment:
For the UPI we'll also want to populate the [`prediction_id`](https://github.com/caraml-dev/universal-prediction-interface/blob/8fbb12ddbc0e7f88d822d36ef685ef6c56475ca3/proto/caraml/upi/v1/upi.proto#L100) with a unique identifier (similar to turing id). Shall we reuse turing id in this case? cc: @krithika369

Reply:
Another doubt I have is to whether store these in the grpc Metadata or in the payload. The drawback of putting it in the payload is we'll need to copy the proto message.